### PR TITLE
HID Shared Memory - Structures + key held/down/up implementation

### PR DESCRIPTION
**Pull request by shinyquagsire23**
Example of usage:
```
// Retrieve only raw left output from Player 1
hidSetControllerLayout(CONTROLLER_PLAYER_1, LAYOUT_LEFT);
while(1)
{
    hidScaninput();

    if (hidKeysHeld(CONTROLLER_PLAYER_1) & KEY_DOWN)
        printf("P1 left controller is held down\n");

    if (hidKeysDown(CONTROLLER_HANDHELD) & KEY_A)
        printf("Handheld pressed A!\n");

    if ((hidMouseButtonsHeld() & MOUSE_LEFT) && hidKeyboardHeld(KBD_W))
        printf("Pyro main detected!\n");

    if (hidKeyboardModifierUp(KBD_MOD_CAPSLOCK))
        printf("caps lock finally released\n");

    if (hidKeysDown(CONTROLLER_HANDHELD) & KEY_TOUCH) {
        touchPosition pos;
        printf("Touches at:\n");
        for (int i = 0; i < hidTouchCount(); i++) {
            hidTouchRead(&pos, i);
            printf("Touch index #%u: %u, %u\n", i, pos.px, pos.py);
        }
    }
}
```

Keyboard keys are handled with the key passed in due to the keyboard keys being held in a 256-bit array, a bit per key based on the HID scancode. Scancodes for keyboards cannot be OR'd together (two separate hidKeyboard* calls OR'd works fine however). Keyboard modifiers can technically be handled the same as controller input, but I had them passed in as arguments for consistency.

Controller input can be retrieved from several different types of layouts, the default layout is `LAYOUT_DEFAULT` which adjusts for controller orientation for Joy-Con halves. This can be overridden to another layout if desired, per-controller.

HID shared memory stores timestamps in number of samples, controller updates are checked for by subtracting the current timestamp from the last timestamp to check if samples have been added. Samples are stored in circular buffers with `latestEntry` indexing the current sample entry. The current behavior follows ctrulib where current samples are cleared and are only set to not-0 if there are updated samples available at the time of calling `hidScanInput`

Touchscreen structures are implemented, I'm still debating exactly how to handle/store held/down/up there since the number of fingers used is variable. More status probing stuff is still incoming as well, but again all structures are implemented for the most part.

- [x] Keys held/down/up for mouse, keyboard, controllers
- [ ] Controller status probing (which controllers are active, what type of controllers are they, how many, what colors are they, what layouts are they using)
- [x] Touchscreen value retrieval, and KEY_TOUCH
- [ ] Activate/deactivate input devices, only probe what is active

### CODE REVIEW

File: nx/source/services/hid.c [hunk 1]
```diff
@@ -61,6 +76,29 @@ void hidExit(void)
     shmemClose(&g_hidSharedmem);
 }
 
+void hidReset(void)
+{
+    // Reset internal state
+    memset(&touchEntry, 0, sizeof(HIDTouchScreenEntry));
+    memset(&mouseEntry, 0, sizeof(HIDMouseEntry));
+    memset(&keyboardEntry, 0, sizeof(HIDKeyboardEntry));
+    memset(&controllerEntries, 0, sizeof(HIDControllerInputEntry));
```
Comment (yellows8) [inline nx/source/services/hid.c]: This is an array not a single entry: remove '&' and use sizeof(controllerEntries).

File: nx/source/services/hid.c [hunk 2]
```diff
@@ -69,6 +107,149 @@ void* hidGetSharedmemAddr(void) {
...
+    memcpy(controllerOld, controllerHeld, sizeof(controllerOld));
+
+    mouseHeld = 0;
+    keyboardModHeld = 0;
+    memset(keyboardHeld, 0, sizeof(keyboardHeld));
+    memset(controllerHeld, 0, sizeof(controllerHeld));
+    memset(&touchEntry, 0, sizeof(HIDTouchScreenEntry));
+    memset(&mouseEntry, 0, sizeof(HIDMouseEntry));
+    memset(&keyboardEntry, 0, sizeof(HIDKeyboardEntry));
+    for (int i = 0; i < 10; i++)
```
Comment (yellows8) [inline nx/source/services/hid.c]: Replace with a memset for the entire array as described above.

File: nx/source/services/hid.c [hunk 3]
```diff
@@ -5,6 +5,19 @@ static Handle g_hidServiceSession = INVALID_HANDLE;
 static Handle g_hidIAppletResource = INVALID_HANDLE;
 static SharedMemory g_hidSharedmem;
 
+static HIDTouchScreenEntry touchEntry;
```
Comment (plutooo) [inline nx/source/services/hid.c]: Can you prefix with g_?

File: nx/source/services/hid.c [hunk 4]
```diff
@@ -61,6 +76,28 @@ void hidExit(void)
     shmemClose(&g_hidSharedmem);
 }
 
+void hidReset(void) {
+    // Reset internal state
+    memset(&touchEntry, 0, sizeof(HIDTouchScreenEntry));
```
Comment (plutooo) [inline nx/source/services/hid.c]: You can turn all of this into an anonymous struct and then just have a single memset-call if you want to

**Comment by fincs** [comment]
How about:
- `u32 hidTouchCount(void)` returning the number of touch points
- `void hidTouchRead(touchPosition* pos, u32 point_id)` for reading a touch point (0 is always known to exist)
- `KEY_TOUCH` pseudo-key that would be added to the key bitmask when `hidTouchCount() != 0`

**Comment by plutooo** [comment]
How to know what controller configuration is active?

File: nx/source/services/hid.c [hunk 5]
```diff
@@ -69,6 +106,135 @@ void* hidGetSharedmemAddr(void) {
...
+    memset(controllerHeld, 0, sizeof(controllerHeld));
+    memset(&touchEntry, 0, sizeof(HIDTouchScreenEntry));
+    memset(&mouseEntry, 0, sizeof(HIDMouseEntry));
+    memset(&keyboardEntry, 0, sizeof(HIDKeyboardEntry));
+    for (int i = 0; i < 10; i++)
+        memset(&controllerEntries[i], 0, sizeof(HIDControllerInputEntry));
+
+    u64 latestTouchEntry = sharedMem->touchscreen.header.latestEntry;
+    HIDTouchScreenEntry *newTouchEntry = &sharedMem->touchscreen.entries[latestTouchEntry];
+    if (newTouchEntry->header.timestamp - touchTimestamp) {
```
Comment (plutooo) [inline nx/source/services/hid.c]: What's going on with this if statement

File: nx/source/services/hid.c [hunk 6]
```diff
@@ -69,6 +106,135 @@ void* hidGetSharedmemAddr(void) {
     return shmemGetAddr(&g_hidSharedmem);
 }
 
+void hidSetControllerLayout(HIDControllerID id, HIDControllerLayoutType layoutType) {
+    controllerLayout[id] = layoutType;
+}
+
+void hidScanInput(void) {
```
Comment (plutooo) [inline nx/source/services/hid.c]: I think we might wanna either split hidScanInput into:

hidScanKeyboard()
hidScanController()
hidScanTouch()
hidScanMouse()
etc...

or have it take a bitfield like this:

hidScanInput(HidScan_Controller | HidScan_Touch)

Comment (plutooo) [reply nx/source/services/hid.c]: For performance reasons

**Comment by yellows8** [comment]
(De)activate service commands should be added too. (http://switchbrew.org/index.php?title=HID_services#hid)

File: nx/source/services/hid.c (hunk 6 above)
Comment (yellows8) [reply nx/source/services/hid.c]: Or only process the currently activated devices?

Comment (fincs) [reply nx/source/services/hid.c]: I think controller state (and touch?) should be unconditionally scanned. The other devices (keyboard, mouse) should only be scanned if they are active/enabled. I don't think `hidScanInput` should be split into multiple functions, or a parameter added to it.

Comment (plutooo) [reply nx/source/services/hid.c]: Why have a special rule? It's better to have the API symmetric for all types of input

File: nx/source/services/hid.c (hunk 5 above)
Comment (fincs) [reply nx/source/services/hid.c]: I think it should be rewritten to `if ((s64)(newTouchEntry->header.timestamp - touchTimestamp) >= 0)`. It is meant to be a basic greater-than check, accounting for the fact that the timestamp overflowing the 64-bit variable is technically possible and valid.

File: nx/source/services/hid.c (hunk 6 above)
Comment (fincs) [reply nx/source/services/hid.c]: But not all types of input will be as commonly used. I think button input should always be available.

File: nx/source/services/hid.c (hunk 5 above)
Comment (plutooo) [reply nx/source/services/hid.c]: Right now the comparison is equivalent to (newTouchEntry->header.timestamp != touchTimestamp) afaict

Comment (fincs) [reply nx/source/services/hid.c]: It is. But of course, incoming touch entries with a previous timestamp would also pass the check... Not sure if we want to assume that will never happen...

File: nx/source/services/hid.c (hunk 6 above)
Comment (plutooo) [reply nx/source/services/hid.c]: Doesn't justify a special rule.

A function should not assume what the user wants to do. It should do what it says, and only that.

Comment (fincs) [reply nx/source/services/hid.c]: Actually, is it possible to disable controller input in the first place? If so, then there wouldn't really be a way to unconditionally disable button input.

I think libnx could keep track of which sources of input are enabled at the moment, and scan those in `hidScanInput`, therefore obviating the need for a special parameter.

File: nx/source/services/hid.c (hunk 5 above)
Comment (shinyquagsire23) [reply nx/source/services/hid.c]: Rollover is unlikely, but I think I'll go ahead and move to a greater-than zero compare for clarity.

File: nx/source/services/hid.c (hunk 6 above)
Comment (shinyquagsire23) [reply nx/source/services/hid.c]: I'm thinking what I'll do is add the activate/deactivate commands @yellows8 mentioned, keep keyboard+mouse off by default at init and then have libnx keep track of and only update button states for devices which are enabled.

I think I can also optimize controllers a bit by checking if the slot has no controller attached, HIDControllerType should have no bits set with controllers detached. I'm not sure if I have anything similar to detect mouse+keyboard though, I'll have to investigate a bit I suppose.

**Comment by shinyquagsire23** [comment]
Went ahead and added a list of things yet to be added, I think at the very least I'd like feature parity with ctrulib w/ touch stuff (maybe minus gyro stuff, need to look into that). Might also look into organizing things into structures some more.

**Comment by plutooo** [comment]
Let's just pull this in

**Pull request merged**